src/server/scene/surface_stack.cpp

```diff
@@ -425,6 +425,108 @@ void ms::SurfaceStack::raise(SurfaceSet const& ss)
     }
 }
 
+void ms::SurfaceStack::swap_z_order(SurfaceSet const& first, SurfaceSet const& second)
+{
+    {
+        RecursiveWriteLock ul(guard);
+        for (auto& layer : surface_layers)
+        {
+            // The goal is to swap the first set with the second set such that their Z-order is swapped.
+
+            // First, find the start and end of the range that we want to swap
+            auto swap_begin = layer.begin();
+            auto swap_end = layer.end();
+            size_t num_first_found = 0;
+            size_t num_second_found = 0;
+            bool first_to_front = false;
+            for (auto it = layer.begin(); it != layer.end(); it++)
+            {
+                // Once we've found them all, then we have the whole range to sort across
+                if (num_first_found == first.size() && num_second_found == second.size())
+                {
+                    swap_end = it;
+                    break;
+                }
+
+                // Find the start position and count how many we've found
+                if (first.count(*it))
+                {
+                    if (!num_first_found && !num_second_found)
+                        swap_begin = it;
+
+                    num_first_found++;
+                }
+                else if (second.count(*it))
+                {
+                    if (!num_first_found && !num_second_found)
+                    {
+                        first_to_front = true;
+                        swap_begin = it;
+                    }
+
+                    num_second_found++;
+                }
+            }
+
+            // Finally, move the to_front items to the front of the group and the to_back to the back of the group
+            auto to_front = first_to_front ? first : second;
+            auto const& to_back = first_to_front ? second : first;
+            std::stable_sort(swap_begin, swap_end, [&](std::weak_ptr<Surface> const& s1, std::weak_ptr<Surface> const& s2)
+            {
+                if (to_front.count(s1))
+                    return to_front.count(s2) == 0;
+                else if (to_front.count(s2))
+                    return false;
+                else if (to_back.count(s1))
+                    return to_back.count(s2) == 0;
```

Comment:
Something is wrong either in the code or in my understanding: If `s1` is in the "to back" partition we should never return true (because `s1` is never strictly before `s2`). But `to_back.count(s2) == 0` can be true.

Reply:
From what I understand, if both s1 and s2 are in `to_back`, then they are in the the correct order (so we return true). This will maintain. The relative order of s1 and s2. If s2 is not in `to_back`, s1 and s2 should be swapped (so we return false)